man/receive.1
.\" $Header$
.\"
.\"  Man page for FUNET-NJE on UNIX system.
.\"
.\"  Written by  Matti Aarnio <mea@finfiles.bitnet> <mea@nic.funet.fi>
.\"
.\"  Date: 27-Dec-1990, 18-Sep-1993, 5-Nov-1993
.\"
.fp 4 CW
.ds ]W Funet-Nje 1 UNIX 1.1
.TH RECEIVE 1
.B "** RAW DRAFT ** RAW DRAFT ** RAW DRAFT ** RAW DRAFT **"
.SH NAME
receive, transfer
\- bitnet incoming file processing programs
.SH SYNOPSIS
.LP
receive
.RB [ "\-n" ]
.RB [ "\-a|\-b" ]
.RB [ "\-asa|\-rawpun|\-bitraw|\-bcat" ]
.RB [ "\-d" ]
.BR [ "\-u "
.I username
.B ]
.RB [ "\-o "
.I outpath
.B ]
.I {filepath|spoolid}
.LP
/usr/lib/huji/transfer
\f4DestUser@DestNode
.I /absolute/file/path
.SH DESCRIPTION
.IX "bitcat"
.IX "bitnet"
.IX "funetnje"
.IX "sendfile"
.IX "netdata"
.IX "ndparse"
.LP
.I Receive
is a relatively simple minded tool for receiving files.
Default usage is to translate incoming
.SM EBCDIC
PUNCH or PRINT records to
.SM ASCII,
and output them with appended newlines.
Trailing spaces may get removed at first, as they are (usually anyway)
unnecessary.
At the 
.SM NETDATA
receiving, no such stripping is done.
.LP
.I Receive
can enter
.I \-d (debug)
mode, and display the contents of the spool file.
At that mode it will display symbolically NJE header contents,
and also analyze
.SM NETDATA
records.
No output file is created in this mode.
.LP
.I Receive
can control output conversion with several keys:
.IP "\fB\-b\fR" 1.0i 0
Turns on enforced binary mode, which outputs the incoming
file as is;
.SM NETDATA
content is analyzed and output as is, and other
.SM PUNCH
modes get also handled in raw manner.
This is \fBdefault\fR mode for a class N
.SM NETDATA
file, and must be explicitly turned off with \fB\-a\fR.
.IP "\fB\-a\fR" 1.0i 0
Turns on enforced
.SM EBCDIC\-ASCII
conversion, even when a
.SM NETDATA
file has class N.
.IP "\fB\-rawpun\fR" 1.0i 0
Outputs nominal size record contents; usually 80 characters per
record
.SM (PUNCH)
without adding anything to the records, nor translating them.
.IP "\fB\-bcat\fR" 1.0i 0
outputs non-translated records in format:

\f4 <char of reclen>  <chars of record>
.IP "\fB\-asa\fR" 1.0i 0
Affects
.SM PRINT
files which contain MCC carriage control codes
.RB "(see " bitspool "(5))"
by translating them to
.SM FORTRAN-style
carriage control.
.LP
.I Receive
outputs to a file whose location is current working directory, and in
non-NETDATA -case the name is formed from spool entry by lowercasing
.IR FileName.FileExt ,
unless
.RB ` \-o ',
is used to override it.
With
.RB ` \-o '
one can specify
.RB ` \- '
to denote
.BR STDOUT .
In case the spool-file contains
.SM NETDATA,
default filename is taken from inside the
.SM "NETDATA INMR02"
control records;
The
.SM INMDSNAM
contains (usually) three entries: \"Disk FileName FileType\",
and they get mapped into UNIX file as:
\"filename.filetype\".
In case the original file is from MVS TSO, and has
.SM INMMEMBR
key defined, resulting UNIX filename is:
\"filename.filetype.membr\".
.LP
Debug output goes to \fISTDERR\fR as usual.
.LP
.I Transfer
modifies file headers changing its destination address so it is possible to
send that file again to NJE network for new destination.
However other security related aspects may not allow all tricks.
(For now origination address in headers will not be changed to point local
user, and thus it is possible to create a file which seems to be traversing
the network in wrong direction.)
.I Transfer
does also requeueing for file transmission.
.LP
.SH OUTPUT FILE NAME MAPPING
User can explicitly define a name for the output file, or let the
.I receive
to invent on its own.
.LP
Default output name is formed from

.LP
When input processing  reaches records to be output, the output file is opened.


.SH OPTIONS
.IP \fInone\fR 1.0i 0
Punch files are output as is, and Print files get
converted from whatever carriage-control format
they use to ASCII-CR/LF/FF format.
.IP "\fB\-asa\fR" 1.0i 0
Print files are output with
.SM FORTRAN-style
carriage controls.  One line per input record.
.IP "\fB\-b\fR" 1.0i 0
Binary output.  See DESCRIPTION on each command.
.IP "\fB\-bitraw\fR" 1.0i 0
Special input format switch for
.IR ndparse .
Globs output from
.I bitcat -b
.IP "\fB\-d\fR" 1.0i 0
Debug output.  To \fISTDERR\fR.
.IP "\fB\-n\fR" 1.0i 0
Original `spool' file should \fBnot\fR be removed after successful
file processing is done.  (Successful == No errors, Netdata file
translated to ASCII using
.I bitcat
might not be useful.)
.IP "\fB\-rawpun\fR" 1.0i 0
Special input format switch for
.IR ndparse .
.LP
.SH ENVIRONMENT
With environment variable
.B EBCDICTBL
user can select alternate EBCDIC <-> ASCII translation
tables.  For more info, see
.BR ebcdictbl (5).
.LP
.SH FILES
.nf
.I /etc/funetnje.cf
.I /etc/funetnje.pid
.RI "File processing exit config file, which is defined in " funetnje.cf.
.fi
.SH SEE ALSO
.BR funetnje (8),
.BR bitspool (5),
.BR send (1),
.BR sendfile (1),
.BR bmail (8),
.BR ebcdictbl (5),
.BR mailify (8),
.BR sysin (8),
.BR qrdr (8),
.BR rdrlist (1)
.SH BUGS
.LP
.I Ndparse
should be able to dig original dates and names from ND file.
.LP
Sometimes
.I ndparse
fails to understand received netdata file.  Subject for further
study.
.LP
.SH DIAGNOSTICS
When any of these programs is fed a file containing erroneous
data (not
.BR funetnje "(8) spool file,)"
these mostly recover with a burp, sometimes these
might even drop a core on pathological input case.
.SH WARNING
With
.I ndparse
it is one time mistake to forget to use
.B \-n
when peeking a file.  That file goes to bit bucket at that
moment if no errors are detected. (rc=0 -> no file...)
.SH AUTHORS
.LP
Yehavi Bourvine, Hebrew University of Yerusalem, Israel
.I <yehavi@hujivms.bitnet>.
.LP
Matti Aarnio, Finnish University and Research Network \- FUNET, Finland
.I <mea@nic.funet.fi> <mea@finfiles.bitnet>.
.LP
Original (VMS oriented package) software by Yehavi Bourvine,
UNIX features, plenty of modifications, man-pages by Matti Aarnio.


ChangeLog
Fri Mar 22 13:48:30 1996  Matti Aarnio  <mea@nic.funet.fi>

	* libsendcmd.c, unix.c:
	    Even more relaxed approach on parametrizing
	    the CMDMAILBOX -- now allows SOCK/FIFO/UDP
	    with fairly flexible (in regards of spaces)
	    syntax..

Mon Nov 13 17:48:52 1995  Matti Aarnio  <mea@nic.funet.fi>

	* libsendcmd.c, unix.c:
	    A bit more relaxed approach on parametrizing
	    CMDMAILBOX config parameter (FIFO/SOCKET case)

	* file_queue.c:
	    Do RSCS-accounting on locally sent files, that
	    wind up to local recipients

	* unix_brcst.c:
	    Re-enter the SysVR4 UTMP weirdos..  Now fixed ?

Fri Aug 25 09:02:55 1995  Matti Aarnio  <mea@nic.funet.fi>

	* mailify.c:
	    Handling error-messages (where "MAIL FROM:" address
	    is "<>", that is, empty) was done improperly.  Now
	    it does flag input as from "channel error" in that
	    case.

Mon Jul 24 14:29:59 1995  Matti Aarnio  <mea@nic.funet.fi>

	* bmail.c, sendfile.c:
	    Sometimes   mktemp()  fails when there are a LOT of
	    files being sent at the same time.  Changed now to
	    more classical  ".MaiL.%05d"-format with  getpid()
	    result as integer data..

Wed Jun 21 22:42:39 1995  Matti Aarnio  <mea@nic.funet.fi>

	* unix_brcst.c, nmr_unix.c:
	    Ok, more than __linux__ is needed to know which
	    machines have <utmp.h> containing extended flags.
	    Now we test  "DEAD_PROCESS" pre-processor define,
	    which existence (propably) will correlate with
	    the existence of utmp's field  ut_type.
	    (SysVr4 feature, I think.)

Mon Jun 19 15:17:55 1995  Matti Aarnio  <mea@nic.funet.fi>

	* send.c, sendfile.c, bmail.c:
	    Some additional tests to trap various forms of bad
	    addresses -- accepts only  user@node format with
	    no '.' in the address, nor either component longer
	    than 8 chars.

	* nmr.c:
	    Made sure not to overwrite allocated string spaces on
	    Msg-queue entry.  (Did cause a crash with it..)

Tue Jun 13 09:13:18 1995  Matti Aarnio  <mea@nic.funet.fi>

	* qrdr.c:
	    Collect directory browsing results to an array at first,
	    then qsort() it with the time-stamp, and finally display
	    them in order. (Latest last)
	    (This is something what Gerald Hanush asked for.)

Mon Jun 12 22:44:00 1995  Matti Aarnio  <mea@nic.funet.fi>

	* file_queue.c:
	    Apparently OSF/1 pipe() return value is the first
	    fd, not "0", when it succeeds.  Nice thing - and not..
	    Altered the test to see if it returns a value under
	    zero, or something positive.  When the return is
	    positive (zero, or larger), fork() and run in
	    child-mode.

	    If a child_pipe exists already ( >= 0 ), close
	    it and mark the new one for processing.

	* libreadcfg.c, read_config, libsendcmd.c, unix.c:
	    Rewrote things so that now the used CMDMAILBOX
	    transport method is chosen at runtime:

	    CMDMAILBOX  F  /path/to/.cmdfifo
	    CMDMAILBOX  f  /path/to/.cmdfifo
	    CMDMAILBOX  S  /path/to/.cmdsocket
	    CMDMAILBOX  U  127.0.0.2 175	# UDP port

	    The "f" is FIFO which is closed, and reopened
	    every time one command is pulled in.

Sun Jun 11 03:12:06 1995  Matti Aarnio  <mea@nic.funet.fi>

	* namesparser.c:
	    Noticed that the data parsing stopped at first ":",
	    which caused premature breaks; observe following entry
	    from BITEARN.NODES -file:
	    	:remark1.BASE(9501) 00/00/00 00:00:00

	* libreadcfg.c, read_config.c, libsendcmd.c, unix.c:
	   Amended "CMDMAILBOX"-data to contain port-number in
	   case UDP-format is used. (That is, have two tokens.)
	   Default value for the port-number is 175.

	* unix_tcp.c read_passive_tcp_connection():
	   Noticed that a case when a socket exists, and a new
	   connection is attempted is handled improperly. The
	   old socket-fd was FORGOTTEN!  Now it NAKs the connection
	   attempt with a complaint that there already is a connection
	   in there...

	   Things will progress from such a situation usually by
	   detecting a closure on the existing stream, but better
	   not to rush ahead of things.

	* bmail.c:
	   Accept "-u" only when calling the program from uids 0 or 1
	   (root and daemon in most common cases -- for BSD-sendmail
	    which wants  bmail executable with protection 2111 (user X,
	    group NJEGRP) )

Sat Jun 10 05:18:27 1995  Matti Aarnio  <mea@nic.funet.fi>

	* Makefile, consts.h, prototypes.h, recv_file.c, io.c,
	  send_file.c, unix.c, protocol.c, unix_tcp.c:

	    - Found one uninitialized pointer, which apparently
	      caused mystic data-structure corruptions (on timer
	      subsystem)

	    - Made mandatory (well, with disabling options) to
	      use NBSTREAM, USE_XMIT_QUEUE, and USE_SOCKOPT.
	      (in consts.h, see also  Makefile's  comments!)

	    - Use  F_WAIT_V_A_BIT  to flag if there is any reason
	      for NOT to send out data at the moment, and instead
	      quit from within  handle_ack()..
	      The F_WAIT_V_A_BIT is set when new data is sent to
	      outbound buffer (primary one, or XMIT QUEUE).
	      The bit is cleared when pending writer finishes
	      with success status.  When the pending-writer is
	      called from  poll_sockets(), immediately after it
	      the  dequeue_xmit_queue() is called, and if there
	      is something, it can cause F_WAIT_V_A_BIT to be
	      set again..)

	    - If the XMIT_QUEUE ever runs out of space, it abort()s!

	    - Tested the beast by sending 80 or so smallish files both
	      ways over the link.  Such setup caused previously link
	      corruption..  Actually the critical phase is when there
	      are flowing NJE headers/trailer, bare data sending works
	      just fine, but when a file ends, EOF block needs an ack
	      to be sent, and unless there the care is used, all
	      queues will overflow..

Thu Jun  8 15:48:14 1995  Matti Aarnio  <mea@nic.funet.fi>

	* namesfilter.c:
	    New option: "-netsoft" to list node software fields.

	* funetnje.texinfo:
	    Some additional chapters

	* unix_tcp.c:

receive.c
/*------------------------------------------------------------------------*
 |		RECEIVE		cat out given file from user queue	  |
 |	For FUNET-NJE by Matti Aarnio  <mea@nic.funet.fi>		  |
 |									  |
 |  Options:								  |
 |	-a		(Translate to ASCII - default, unless CLASS N)	  |
 |	-b		(binary  length(byte)+contents  format)		  |
 |      -d              (debug dump of spool file contents)		  |
 |	-n		(Don't delete upon successfull reception of ND)	  |
 |	-o filepath	(Output file name)				  |
 |	-asa		(print files with 'Fortran style' carriage ctrl)  |
 |	-rawpun		(Plain 80/132 bytes per record fixed size)	  |
 |	-bitraw
 |	-bcat
 |	-u username
 |	filepath	(point to file to search)			  |
 |									  |
 |									  |
 *------------------------------------------------------------------------*/

/*
   Thoughts about developement:
	- can be invoced with multiple names to default several functions.
	- need to learn about "CARD" format (Cornell ?)
	- need to learn about NETDATA MAIL format (PROFS NOTEs)
	- A generic output channel/procedure arg  in place of outfile
	  (to make this be an usable base for writing a single-module
	   mailify)

 */


#include "headers.h"
#include "prototypes.h"
#include "clientutils.h"
#include "ndlib.h"
#include <utime.h>

char LOCAL_NAME   [10];
char BITNET_QUEUE [80];
char LOG_FILE     [80] = "-";

extern char DefaultSpoolDir[80];

int LogLevel = 2;
FILE *LogFd = NULL;

extern	FILE	*dump_header __(( char const *path, const int debugdump, int *binary ));
extern	void	debug_headers();
extern	void	pad_to_size();
extern	void	dump_punchline __((struct ndparam *ndp, unsigned char *buf, int len, int reclen, const int style));
extern	void	dump_printmccline __((struct ndparam *ndp, unsigned char *buf, int len, int reclen, const int style));
extern	void	dump_printasaline __((struct ndparam *ndp, unsigned char SRCB, unsigned char *buf, int len, int reclen, const int style));
extern	char	*Post;
extern	int	dump_file __(( char const *path, char const *outpath, int binary, const int dumpstyle, const int debugdump ));
extern	void	usage __(( void ));

char *pname = "<RECEIVE>";

struct ndparam ndparam;	/* For NETDATA parser */


int
main(argc,argv)
     int argc;
     char *argv[];
{
	int debugdump = 0;
	int dumpstyle = 1;	/* default style: ASCII output */
	int binary = 0;
	int rc = 0;
	int nodelete = 0;
	int rawpun = 0;
	int rawin = 0;
	int bitraw;
	const	char *path;
	char	*outpath = NULL;
	char	HomeDir[256];
	char	WhoAmI[10];
	char	WhoAmIl[10];	/* Lowercase copy.. */
	char	StudyDir[256];
	
	strcpy(LOG_FILE,"-");
	WhoAmI[0] = 0;

	pname = strrchr(*argv,'/');
	if (pname) ++pname;
	else pname = *argv;

	if(argc < 2) usage();

	read_configuration();
	read_etable();

	++argv;
	while (*argv) {
	  if (strcmp(*argv,"-n")==0) {
	    nodelete = 1;
	  } else if (strcmp(*argv,"-a")==0) {
	    binary = -1;
	  } else if (strcmp("-asa",*argv)==0) {
	    if (dumpstyle != 1) usage();
	    dumpstyle = 2;	/* To ascii, show ASA codes */
	  } else if (strcmp(*argv,"-b")==0) {
	    if (dumpstyle != 1) usage();
	    dumpstyle = 0;	/* binary style... */
	    binary = 1;
	  } else if (strcmp(*argv,"-bcat")==0) {
	    binary = 2;
	  } else if (strcmp(*argv,"-bitraw")==0) {
	    rawpun = 1;
	    bitraw = 1;
	  } else if (strcmp("-d",*argv)==0) {
	    debugdump = 1;
	  } else if (strcmp(*argv,"-o")==0) {
	    ++argv;
	    if (!*argv) usage();
	    outpath = *argv;
	  } else if (strcmp(*argv,"-u")==0) {
	    ++argv;
	    if (!*argv) usage();
	    strncpy(WhoAmI,*argv,8);
	  } else if (strcmp(*argv,"-rawin")==0) {
	    rawin = 1;
	  } else if (strcmp(*argv,"-rawpun")==0) {
	    if (dumpstyle != 1) usage();
	    dumpstyle = -1;	/* Binary style - sort of */
	  } else if (**argv == '-')
	    usage();
	  else
	    break;
	  ++argv;
	}
	path = *argv;
	if (path == NULL && rawin)
	  path = "-";
	if (path == NULL)
	  usage();
	if (access(path,R_OK|F_OK) != 0) {
	  /* Well, not a local file, try expand it with
	     default spooldir, etc.  Thus allow using `spoolid'
	     argument as is. */

	  struct passwd *passwdent;
	  char *pathend;

	  if (*WhoAmI == 0) {
	    char *p = getenv("LOGNAME"); /* BSD/SunOS/... alike */
	    WhoAmI[8] = 0;
	    if (!p) {
	      p = getenv("USER"); /* SysV */
	      if (!p) {
		passwdent = getpwuid(getuid());
		if (passwdent)
		  strncpy(WhoAmI,passwdent->pw_name,8);
		else {
		  fprintf(stderr,"RECEIVE: Can't figure out who I am.\n");
		  exit(2);
		}
	      } else
		strncpy(WhoAmI,p,8);
	    } else
	      strncpy(WhoAmI,p,8);
	  }
	  strcpy(WhoAmIl,WhoAmI);
	  upperstr(WhoAmI);

	  strcpy(StudyDir,DefaultSpoolDir);
	  if ((passwdent = getpwnam( WhoAmIl )) != NULL) {
	    strcpy(HomeDir,passwdent->pw_dir);
	  } else {
	    logger(1,"RECEIVE: Can't figure home dir of `%s' -- PseudoUser ?\n",
		   WhoAmI);
	    strcpy(HomeDir,"/");
	  }
	  ExpandHomeDir(DefaultSpoolDir,HomeDir,WhoAmI,StudyDir);

	  pathend = StudyDir+strlen(StudyDir);
	  strcpy(pathend,*argv);

	  if (access(StudyDir,R_OK|F_OK)!=0) {
	    /* Doesn't exist with that name, see if it is numeric.. */
	    int spoolid = atoi(*argv);
	    if (spoolid >= 1 && spoolid <= 9900)
	      sprintf(pathend,"%04d",spoolid);
	    /* Fails, or not, dump_file() will report.. */
	  }

	  path = StudyDir;
	}

	rc = dump_file(path,outpath,binary,dumpstyle,debugdump);

	if (!nodelete && !rc)
	  unlink(path);

	return rc;
}

extern char FileName[], FileExt[];


int
dump_file(path,outpath,binary,dumpstyle,debugdump)
     char const *path;
     char const *outpath;
     int binary;
     const int dumpstyle;
     const int debugdump;
{
	unsigned char	line[1005];
	FILE	*fd;
	int NewSize;
	int recs = 0;
	int linelen;
	int is_netdata = 0;
	int known_format = 0;
	int rc = 0;
	char defaultoutname[33];
	struct utimbuf times;

	/* Setup netdata parser */
	ndparam.outfile = NULL;
	ndparam.outname = outpath;
	ndparam.defoutname = defaultoutname;
	ndparam.profsnote = 0;
	ndparam.ackwanted = 0;
	ndparam.outproc  = NULL;
	ndparam.DSNAM[0] = 0;
	ndparam.ftimet   = 0;
	ndparam.ctimet   = 0;
	ndparam.RecordsCount = 0;

	Post = NULL;

	if (strcmp(path,"-")==0) {
	  fd = stdin;
	  strcpy(FileName,"standard");
	  strcpy(FileExt, "input");
	} else
	  fd = dump_header(path,debugdump,&binary);
	if (!fd) return 3;

	if (binary < 0) binary = 0; /* ASCII xlate */

	sprintf(defaultoutname,"%s.%s",FileName,FileExt);
	lowerstr(defaultoutname);


	while ((NewSize = Uread(line,sizeof line, fd)) != 0) {
	  ++recs;
	  if (debugdump) {
	    debug_headers(line,NewSize);
	    trace(line,NewSize,1);
	  }
	  switch (*line) {

	    case CC_NO_SRCB:
		/* Can also be NETDATA; test if format is not yet know,
		   and if so, test for ND.  If it is ND, proceed with
		   handling it.. */

		linelen = *(line+1);

		if (!known_format) {
		  /* XX: Knows only about NETDATA, how about
		     CORNELL CARD, et.al ?				*/
		  if (memcmp("\311\325\324\331\360\361",line+4,6)==0)
		    is_netdata = 1; /* Detected INMR01 at the begin
				       of a record */
		  known_format = 1;
		}

		if (is_netdata) {
		  pad_to_size(line+2,NewSize-2,linelen,E_SP);
		  if ((rc = parse_netdata(line+2,linelen,
					  &ndparam,
					  binary,debugdump))) {
		    if (rc == -1)
		      rc = 0;	/* EOF, all ok */
		    else
		      printf("error from parse_netdata() rc=%d, recs=%d\n",
			     rc,recs);
		  }
		} else {
		  if (debugdump) break;

		  if (ndparam.outfile == NULL) {
		    if (outpath == NULL)
		      outpath = defaultoutname;
		    if (strcmp("-",outpath)==0) ndparam.outfile = stdout;
		    else if ((ndparam.outfile = fopen(outpath,"w")) == NULL) {
		      fprintf(stderr,"RECEIVE: outpath `%s' isn't openable.  Maybe you should try -o option ?\n",outpath);
		      return 2;
		    }
		    setbuf(ndparam.outfile,NULL);
		  }

		  /* ----- dump punch line raw contents ----- */
		  if (dumpstyle <= 0) { /* Raw style */
		    pad_to_size(line+2,NewSize-2,linelen,E_SP);
		    dump_punchline(&ndparam,line+1,linelen,linelen,dumpstyle);
		  } else	/* ASCII conversion... */
		    dump_punchline(&ndparam,line+1,linelen,NewSize-2,dumpstyle);
		}
		break;

	    case CC_MAC_SRCB:

		if (debugdump) break;
		
		if (ndparam.outfile == NULL) {
		  if (outpath == NULL)
		    outpath = defaultoutname;
		  if (strcmp("-",outpath)==0) ndparam.outfile = stdout;
		  else if ((ndparam.outfile = fopen(outpath,"w")) == NULL) {
		    fprintf(stderr,"RECEIVE: outpath `%s' isn't openable.  Maybe you should try -o option ?\n",outpath);
		    return 2;
		  }
		  setbuf(ndparam.outfile,NULL);
		}

		known_format = 1; /* Won't be NETDATA -- I hope.. */

		linelen = *(line+1);
		if (dumpstyle <= 0) { /* Raw style */
		  pad_to_size(line+2,NewSize-2,linelen,E_SP);
		  dump_printmccline(&ndparam,line+1,linelen,linelen,dumpstyle);
		} else	/* ASCII conversion... */
		  dump_printmccline(&ndparam,line+1,linelen,NewSize-3,dumpstyle);
		break;

	    case CC_ASA_SRCB:
	    case CC_CPDS_SRCB:

		if (debugdump) break;

		if (ndparam.outfile == NULL) {
		  if (outpath == NULL)
		    outpath = defaultoutname;
		  if (strcmp("-",outpath)==0) ndparam.outfile = stdout;
		  else if ((ndparam.outfile = fopen(outpath,"w")) == NULL) {
		    fprintf(stderr,"RECEIVE: outpath `%s' isn't openable.  Maybe you should try -o option ?\n",outpath);
		    return 2;
		  }
		  setbuf(ndparam.outfile,NULL);
		}

		/*  ----- dump print line raw contents ----- */
		known_format = 1; /* Won't be NETDATA -- I hope.. */
		linelen = *(line+1);
		if (dumpstyle <= 0) { /* Raw style */
		  pad_to_size(line+2,NewSize-2,linelen,E_SP);
		  dump_printasaline(&ndparam,*line, line+1,linelen,linelen,
				    dumpstyle);
		} else		/* ASCII conversion...	*/
		  dump_printasaline(&ndparam,*line, line+1,linelen,NewSize-3,
				    dumpstyle);
		break;

	    default:		/* All others		*/
		/* Surprise: Forget it */
		break;
	    }
	}

	/* Add the trailing newline, if need be.. */
	if (dumpstyle == 1 && Post && !ndparam.outproc && ndparam.outfile)
	  fputs(Post,ndparam.outfile);
	Post = NULL;

	if (!ndparam.outproc && ndparam.outfile)
	  fclose(ndparam.outfile);
	fclose(fd);

	if (!ndparam.outproc && ndparam.outfile) {
	  /* Pick the original file date -- from INMCREAT, if present,
	     or from INMFTIME, if present.  If neither is present, don't
	     change times.. */
	  times.actime  = time(NULL);
	  times.modtime = ndparam.ctimet;

	  if (times.modtime == 0 || times.modtime == (time_t) -1)
	    times.modtime = ndparam.ftimet;

	  if (times.modtime != 0 && times.modtime != (time_t) -1)
	    utime(ndparam.outname,&times);

	  /* printf("NETDATA: CTIMES='%s', ctimet=%d, FTIMES=%d, ftimet=%d, time=%s",
	     ndparam.CTIMES,ndparam.ctimet,ndparam.FTIMES,ndparam.ftimet,
	     ctime(&ndparam.ftime)); */
	}

	if (debugdump) {
	  debug_headers("\0\0\0\0\0\0\0\0",8);
	  printf("\n%d records.\n",recs);
	}

	return rc;

}

volatile void
bug_check(string)
char const      *string;
{
	fprintf(stderr, "RECEIVE: Bug check: %s\n", string);
        exit(0);
}


void
usage()
{
	fprintf(stderr,"%s:  [-d] [-b | -rawpun] filename\n",pname);
	fprintf(stderr,"%s:  -a       output converted to ASCII (default).\n",pname);
	fprintf(stderr,"%s:  -asa     output contains ASA carriage control.\n",pname);
	fprintf(stderr,"%s:  -b       output raw (binary) of bitcat style.\n",pname);
	fprintf(stderr,"%s:  -d       debug mode\n",pname);
	fprintf(stderr,"%s:  -n       don't delete the spool file on successfull processing.\n",pname);
	fprintf(stderr,"%s:  -o file  output to named filed.\n",pname);
	fprintf(stderr,"%s:  -rawpun  output raw nominal size record contents. 80 for punch, etc.\n",pname);
	fprintf(stderr,"%s: AND OTHERS -- to be defined/documented!\n",pname);
	exit(1);
}
